lib/ui/volatile_path_tracker.h

```diff
@@ -66,14 +67,10 @@ class VolatilePathTracker {
 
  private:
   fml::RefPtr<fml::TaskRunner> ui_task_runner_;
-  std::atomic_bool needs_drain_ = false;
-  std::mutex paths_to_remove_mutex_;
-  std::deque<std::shared_ptr<TrackedPath>> paths_to_remove_;
-  std::set<std::shared_ptr<TrackedPath>> paths_;
+  std::mutex paths_mutex_;
+  std::vector<std::shared_ptr<TrackedPath>> paths_;
```

Comment:
How about just maintain a `set` of `std::weak_ptr`s to TrackedPath? The ones you can `lock` successfully in the `OnFrame` have by definition survived and you can bump up the `frame_count` on them and set volatility appropriately. The ones you cannot lock can be discarded from the set. You don't need an explicit `Erase` call in that case. The Insert can be renamed to `Track`.

Reply:
I like this, but I think to make it work we'll have to make `ReleaseDartWrapper` non-const.

Reply:
(so that I can call `shared_ptr::reset` from it)

Reply:
Hm. That has some weird side effects of making subsequent calls to `path()` unsafe. I'll see if instead I can update the tests that are relying on this behavior to track things.

Reply:
Yeah, totally. I don't know why that is `const` in the first place. I think I wrote that in tonic and the only reason it was const was because `AddRef` and `Release` were const in `fml::RefCountedThreadSafeBase`. But a pure virtual interface specifying the constness of that call is unreasonable especially because ref counts are being modified. It's just that it was written specifically for FML which was const to being with.

Reply:
If that is the only blocker, I think removing the const-ness requirement is the way to go. That would avoid the extra call to Erase and avoid the need to have a mutex. Ref count bumps and the lock attempts are atomic and thread safe.

Reply:
I just read https://github.com/flutter/engine/pull/30299/files#r768163175 and I am not sure I follow which `path()` is unsafe.

Reply:
I had to tweak the test that validated GC did the right thing a bit - but PTAL, it's up to date now.

Reply:
And since weak_ptr is thread safe I'm not in need of a mutex at all anymore

Reply:
@chinmaygarde - it was a version of this patch that I didn't end up committing because tests were segfaulting - that was me thinking out loud, sorry. Basically I was trying to call `reset` on the tracked path on GC signal, which meant if someone held a reference beyond that and tried to call `path()` they'd segfault.